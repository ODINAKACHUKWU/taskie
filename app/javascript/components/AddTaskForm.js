import React, { useState } from "react";
import { useHistory } from "react-router-dom";
import axios from "axios";

function AddTaskForm() {
  const history = useHistory();
  const [title, setTitle] = useState("");
  const [description, setDescription] = useState("");
  const [avatarUrl, setAvatarUrl] = useState("");

  const handleSubmit = (e) => {
    e.preventDefault();
    const token = document.querySelector("[name=csrf-token]").content;
    axios.defaults.headers.common["X-CSRF-TOKEN"] = token;
    axios
      .post("/api/v1/tasks", { title, description, avatarUrl })
      .then((res) => history.push("/"))
      .catch((err) => console.log(err));
  };

  return (
    <form id="add-task-form" onSubmit={handleSubmit}>
      <div>
        <input
          id="title-input"
          type="text"
          placeholder="Task Title"
          value={title}
          onChange={(e) => setTitle(e.target.value)}
          required
        />
      </div>
      <div>
        <input
          id="description-input"
          type="text"
          placeholder="Task Description"
          value={description}
          onChange={(e) => setDescription(e.target.value)}
        />
      </div>
      <div>
        <input
          id="avatar-url-input"
          type="text"
          placeholder="Avatar URL"
          value={avatarUrl}
          onChange={(e) => setAvatarUrl(e.target.value)}
        />
      </div>
      <div>
        <button id="add-button" type="submit">
          Add
        </button>
      </div>
    </form>
  );
}

export default AddTaskForm;
