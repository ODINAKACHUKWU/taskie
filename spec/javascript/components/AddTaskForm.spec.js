import "jsdom-global/register";
import React from "react";
import { shallow, mount } from "enzyme";
import toJSON from "enzyme-to-json";
import AddTaskForm from "../../../app/javascript/components/AddTaskForm";

const api = {
  addTask: () => ({
    title: "Title",
    description: "Description",
    avatar_url: "",
  }),
};

const simulateChangeOnInput = (wrapper, inputSelector, newValue) => {
  const input = wrapper.find(inputSelector);
  input.simulate("change", {
    target: { value: newValue },
  });

  return wrapper.find(inputSelector);
};

describe("rendering AddTaskForm component", () => {
  it("renders AddTaskForm component without crashing", () => {
    shallow(<AddTaskForm />);
  });

  it("renders AddTaskForm as a form element", () => {
    const wrapper = shallow(<AddTaskForm />);

    expect(wrapper.find("#add-task-form").type()).toEqual("form");
  });

  it("renders input fields without crashing", () => {
    const wrapper = mount(<AddTaskForm />);
    const inputs = wrapper.find("input");

    expect(inputs.length).toEqual(3);
  });

  it("renders Add button", () => {
    const wrapper = shallow(<AddTaskForm />);
    const label = wrapper.find("#add-button");

    expect(label.type()).toEqual("button");
    expect(label.text()).toEqual("Add");
  });
});

describe("filling form", () => {
  beforeEach(() => jest.resetAllMocks());

  it("lets user fill out the form with values", () => {
    const wrapper = shallow(<AddTaskForm />);

    const titleInput = simulateChangeOnInput(wrapper, "#title-input", "Title");
    const descriptionInput = simulateChangeOnInput(
      wrapper,
      "#description-input",
      "Description"
    );
    const avatarInput = simulateChangeOnInput(
      wrapper,
      "#avatar-url-input",
      "Avatar URL"
    );

    expect(titleInput.props().value).toEqual("Title");
    expect(descriptionInput.props().value).toEqual("Description");
    expect(avatarInput.props().value).toEqual("Avatar URL");
  });

  it("it submits the form to the API", () => {
    document.body.innerHTML =
      "<input name='csrf-token' value='gibhayuvh7wbihh' />";
    jest.spyOn(api, "addTask").mockImplementation(() =>
      Promise.resolve({
        title: "Title",
        description: "Description",
        avatar_url: "",
      })
    );

    const wrapper = mount(<AddTaskForm />);
    const titleInput = simulateChangeOnInput(wrapper, "#title-input", "Title");
    const descriptionInput = simulateChangeOnInput(
      wrapper,
      "#description-input",
      "Description"
    );
    const avatarInput = simulateChangeOnInput(
      wrapper,
      "#avatar-url-input",
      "Avatar URL"
    );

    wrapper.find("form#add-task-form").simulate("submit", {
      preventDefault: () => {},
    });

    expect(api.addTask).toHaveBeenCalled();
  });
});

describe("snapshot", () => {
  it("AddTaskForm snapshot", () => {
    const tree = shallow(<AddTaskForm />);

    expect(toJSON(tree)).toMatchSnapshot();
  });
});
