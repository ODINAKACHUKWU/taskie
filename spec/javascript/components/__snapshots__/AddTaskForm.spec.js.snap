// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot AddTaskForm snapshot 1`] = `
<form
  id="add-task-form"
  onSubmit={[Function]}
>
  <div>
    <input
      id="title-input"
      onChange={[Function]}
      placeholder="Task Title"
      required={true}
      type="text"
      value=""
    />
  </div>
  <div>
    <input
      id="description-input"
      onChange={[Function]}
      placeholder="Task Description"
      type="text"
      value=""
    />
  </div>
  <div>
    <input
      id="avatar-url-input"
      onChange={[Function]}
      placeholder="Avatar URL"
      type="text"
      value=""
    />
  </div>
  <div>
    <button
      id="add-button"
      type="submit"
    >
      Add
    </button>
  </div>
</form>
`;
